### PR TITLE
Build The Operator and Tests with Docker

### PR DESCRIPTION
**Pull request by chatton**
### All Submissions:

* [ ] Have you opened an Issue before filing this PR?
* [X] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

This PR is in favour of #18.

This PR introduces several different Dockerfiles

* Dockerfile.unittest will run `go test` on `pkg`
* Dockerfile.e2e ultimately runs `operator-sdk test --local` and should be deployed the Kind cluster which is created on evergreen (or locally during development)
* Dockerfile.operator is a multi stage Dockerfile which builds the operator during the first stage, while the second stage is almost identical to the current operator Dockerfile, with the exception we are taking the binary from within the multi stage build.

This has the benefit of not decoupling the go build/run/test environments from our CI, and will result in reproducable environments during development.

### CODE REVIEW

File: deploy/operator.yaml [hunk 1]
```diff
@@ -16,7 +16,7 @@ spec:
       containers:
         - name: mongodb-kubernetes-operator
           # Replace this with the built image name
-          image: REPLACE_IMAGE
+          image: quay.io/chatton/mongodb-kubernetes-operator
```
Comment (chatton) [inline deploy/operator.yaml]: these values are temporary during development

File: docker/Dockerfile.operator [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+# copy the operator binary
+COPY --from=builder /go/build/_output/bin/mongodb-kubernetes-operator ${OPERATOR}
+
+COPY --from=builder /go/build/bin /usr/local/bin
+RUN  /usr/local/bin/user_setup
+
+ENTRYPOINT ["/usr/local/bin/entrypoint"]
+
+USER ${USER_UID}
```
Comment (chatton) [inline docker/Dockerfile.operator]: this entirely taken from the auto generated Dockerfile created by `operator-sdk new` with `--from=builder ` being the only significant difference

File: scripts/ci/run_test.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+# This is a temporary fix to correct the KUBECONFIG file that gets mounted.
+# When the KUBECONFIG file is created by kind, it points to localhost. When inside
+# the cluster, we need to set this value to be .spec.clusterIP instead
+# See: https://github.com/operator-framework/operator-sdk/issues/2618
+KUBERNETES_SERVICE_HOST="$(kubectl get service kubernetes -o jsonpath='{.spec.clusterIP }')"
+temp=$(mktemp)
+cat ${KUBECONFIG} | sed "s/server: https.*/server: https:\/\/${KUBERNETES_SERVICE_HOST}/g" > ${temp}
+contents="$(cat ${temp})"
+kubectl create cm kube-config --from-literal=kubeconfig="${contents}"
+rm ${temp}
```
Comment (chatton) [inline scripts/ci/run_test.sh]: this needs to go ASAP

File: docker/Dockerfile.e2e [hunk 4]
```diff
@@ -0,0 +1,16 @@
+FROM golang
+
+ENV OEPRATOR_SDK_VERSION v0.15.2
```
Comment (rodrigovalin) [inline docker/Dockerfile.e2e]: OEPRATOR ;)

Comment (chatton) [reply docker/Dockerfile.e2e]: 🤦‍♂

File: pkg/kube/client/mocked_manager.go [hunk 5]
```diff
@@ -96,3 +98,13 @@ func (m *MockedManager) GetRESTMapper() meta.RESTMapper {
...
+
+// AddHealthzCheck allows you to add Healthz checker
+func (m *MockedManager) AddHealthzCheck(name string, check healthz.Checker) error {
+	return nil
+}
+
+// AddReadyzCheck allows you to add Readyz checker
+func (m *MockedManager) AddReadyzCheck(name string, check healthz.Checker) error {
+	return nil
+}
```
Comment (rodrigovalin) [inline pkg/kube/client/mocked_manager.go]: can we delete this?

File: scripts/ci/run_test.sh [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+# This is a temporary fix to correct the KUBECONFIG file that gets mounted.
+# When the KUBECONFIG file is created by kind, it points to localhost. When inside
+# the cluster, we need to set this value to be .spec.clusterIP instead
+# See: https://github.com/operator-framework/operator-sdk/issues/2618
+KUBERNETES_SERVICE_HOST="$(kubectl get service kubernetes -o jsonpath='{.spec.clusterIP }')"
+temp=$(mktemp)
+cat ${KUBECONFIG} | sed "s/server: https.*/server: https:\/\/${KUBERNETES_SERVICE_HOST}/g" > ${temp}
+contents="$(cat ${temp})"
+kubectl create cm kube-config --from-literal=kubeconfig="${contents}"
```
Comment (rodrigovalin) [inline scripts/ci/run_test.sh]: you can do `--from-file=kubeconfig` (file needs to be called kubeconfig)

**Pull request merged**